File: source/native/src/MugenExpression.cpp
#include "MugenExpression.hpp"
#include "expression/Context.hpp"

Parser MugenExpression::parser = Parser();

void MugenExpression::_init() {
	// pass
}

void MugenExpression::_register_methods() {
	register_method("parse", &MugenExpression::parse);
    register_method("execute", &MugenExpression::execute);
    register_method("get_error_text", &MugenExpression::get_error_text);
    register_method("to_string", &MugenExpression::to_string);
    register_method("has_error", &MugenExpression::has_error);
}

void MugenExpression::parse(String expression_)
{
    error = "";

    expression = parser.parse(expression_.utf8().get_data());

    if (parser.error.size() > 0) {
        error = parser.error;
    }
}

Variant MugenExpression::execute(Variant context_)
{
    error = "";

    shared_ptr<Context> context(new ObjectContext(context_));

    if (expression == NULL) {
        error = "Expression not parsed";
        return Variant();
    }

    Value result = expression->evaluate(context.get());

    if (result.isBottom()) {
        error = "Expression returned bottom";
    }

    return result;
}

String MugenExpression::get_error_text() const
{
    return error.c_str();
}

bool MugenExpression::has_error() const
{
    return error.size() > 0;
}

String MugenExpression::to_string() const
{
    return String(expression->toString().c_str());
}

File: source/native/src/expressions/mugen_expression.rs
use crate::expressions::context::Context;
use crate::expressions::expression::Expression;
use crate::expressions::parser::{parse, ParserState};
use gdnative::prelude::*;
use std::cell::RefCell;
use std::rc::Rc;

#[derive(NativeClass)]
#[inherit(Reference)]
pub struct MugenExpression {
    error: String,
    state_ref: Rc<RefCell<ParserState>>,
    expression: Rc<Expression>,
}

#[methods]
impl MugenExpression {
    /// The "constructor" of the class.
    fn new(_owner: &Reference) -> Self {
        MugenExpression {
            error: "".to_string(),
            state_ref: Rc::new(RefCell::new(ParserState::new())),
            expression: Rc::new(Expression::BottomExpression),
        }
    }

    #[method]
    fn parse(&mut self, text: String) {
        self.expression = parse(Rc::clone(&self.state_ref), text);
        self.error = self.state_ref.borrow().error.to_string();
    }

    #[method]
    fn execute(&self, context: Variant) -> Variant {
        let context_object = Context::new(context);

        self.expression.evaluate(&context_object).to_variant()
    }

    #[method]
    fn get_error_text(&self) -> String {
        self.error.to_string()
    }

    #[method]
    fn has_error(&self) -> bool {
        !self.error.is_empty()
    }

    #[method]
    fn to_string(&self) -> String {
        self.expression.to_string()
    }
}
